File: plugin/src/test/java/com/stratio/cassandra/lucene/util/DateParserTest.java
/*
 * Copyright (C) 2014 Stratio (http://stratio.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.stratio.cassandra.lucene.util;

import com.stratio.cassandra.lucene.IndexException;
import com.stratio.cassandra.lucene.schema.column.Column;
import org.apache.cassandra.db.marshal.Int32Type;
import org.apache.cassandra.db.marshal.SimpleDateType;
import org.junit.Test;

import java.nio.ByteBuffer;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.fail;

/**
 * @author Eduardo Alonso  {@literal <eduardoalonso@stratio.com>}
 */
public class DateParserTest {

    @Test
    public void testParseNullDate() {
        DateParser dateParser = new DateParser("yyyy/MM/dd");
        Date date = null;
        Date dateOut = dateParser.parse(date);
        assertEquals("Date null parsed by DateParser(\"yyyy/MM/dd\") must return null", null, dateOut);
    }

    @Test
    public void testParseNullObject() {
        DateParser dateParser = new DateParser("yyyy/MM/dd");
        Object date = null;
        Date dateOut = dateParser.parse(date);
        assertEquals("Object null parsed by DateParser(\"yyyy/MM/dd\") must return null", null, dateOut);
    }

    @Test
    public void testParseNullLong() {
        DateParser dateParser = new DateParser("yyyy/MM/dd");
        Long date = null;
        Date dateOut = dateParser.parse(date);
        assertEquals("Long null parsed by DateParser(\"yyyy/MM/dd\") must return null", null, dateOut);
    }

    @Test
    public void testParseValidDate() throws ParseException {
        String pattern = "yyyy/MM/dd";
        DateParser dateParser = new DateParser(pattern);

        String dateString = "2015/11/03";
        SimpleDateFormat simpleDateFormat = new SimpleDateFormat(pattern);

        Date date = simpleDateFormat.parse(dateString);

        Date dateOut = dateParser.parse(date);
        assertEquals("Date parsed by DateParser(\"yyyy/MM/dd\") wrong parsed",
                     dateString,
                     simpleDateFormat.format(dateOut));

    }

    @Test
    public void testParseValidObject() {

        String pattern = "yyyy/MM/dd";
        SimpleDateFormat simpleDateFormat = new SimpleDateFormat(pattern);

        DateParser dateParser = new DateParser(pattern);

        String dateString = "2015/11/03";
        Date dateOut = dateParser.parse(dateString);

        assertEquals("String parsed by DateParser(\"yyyy/MM/dd\") wrong parsed",
                     dateString,
                     simpleDateFormat.format(dateOut));
    }

    @Test
    public void testParseValidLong() {

        DateParser dateParser = new DateParser(null);
        Long dateLong = 20151103l;
        Date dateOut = dateParser.parse(dateLong);

        assertEquals("Long parsed by DateParser(\"yyyy/MM/dd\") wrong parsed",
                     dateLong.toString(),
                     Long.toString(dateOut.getTime()));
    }

    @Test(expected = IndexException.class)
    public void testParseInvalidObject() throws ParseException {
        String pattern = "yyyyMMdd";
        DateParser dateParser = new DateParser(pattern);

        String dateString = "2015/11/03";

        dateParser.parse(dateString);
        fail("DateParser(" + pattern + ").parse(" + dateString + ") Must generate IndexException and does not do it");

    }

    @Test(expected = IndexException.class)
    public void testParseInvalidNegativeLong() {
        String pattern = "yyyyMMdd";
        DateParser dateParser = new DateParser(pattern);

        Long dateLong = -20152345l;//invalid long not parseable
        dateParser.parse(dateLong);
        fail("DateParser(" +
             pattern +
             ").parse(" +
             dateLong.toString() +
             ") Must generate IndexException and does not do it");

    }

    @Test
    public void testValidTimestampString() {
        String pattern = "timestamp";
        DateParser dateParser = new DateParser(pattern);
        String dateString = "2635421542648178234";
        Date date = dateParser.parse(dateString);

        Date dateToCompare = new Date(Long.parseLong(dateString));
        assertEquals("DateParser(" + pattern + ").parse(" + dateString + ") wrong parsed.", dateToCompare, date);
    }

    @Test
    public void testValidTimestampLong() {
        String pattern = "timestamp";
        DateParser dateParser = new DateParser(pattern);
        Long dateLong = 2635421542648178234l;
        Date date = dateParser.parse(dateLong);

        Date dateToCompare = new Date(dateLong);
        assertEquals(String.format("DateParser(%s).parse(%s) wrong parsed.", pattern, dateLong), dateToCompare, date);
    }

    @Test(expected = IndexException.class)
    public void testInvalidTimestampString() {
        String pattern = "timestamp";
        DateParser dateParser = new DateParser(pattern);

        String dateString = "2015/03/02";
        dateParser.parse(dateString);
        fail(String.format("DateParser(%s).parse(%s) Must generate IndexException and does not do it",
                           pattern,
                           dateString));
    }

    @Test
    public void testParseColumnSimpleDateSamePattern() {
        String pattern = "yyyy-MM-dd";
        DateParser parser = new DateParser(pattern);
        ByteBuffer bb = SimpleDateType.instance.fromString("2015-10-10");
        Column<Integer> column = Column.builder("date").decomposedValue(bb, SimpleDateType.instance);
        Date actualDate = parser.parse(column);
        Date expectedDate = parser.parse("2015-10-10");
        assertEquals("Date parser fails with SimpleDateType and default pattern", expectedDate, actualDate);
    }

    @Test
    public void testParseColumnSimpleDateDifferentPattern() {
        String pattern = "yyyy/MM/dd";
        DateParser parser = new DateParser(pattern);
        ByteBuffer bb = SimpleDateType.instance.fromString("2015-10-10");
        Column<Integer> column = Column.builder("date").decomposedValue(bb, SimpleDateType.instance);
        Date actualDate = parser.parse(column);
        Date expectedDate = parser.parse("2015/10/10");
        assertEquals("Date parser fails with SimpleDateType and different pattern", expectedDate, actualDate);
    }

    @Test
    public void testParseColumnSimpleDateTimestampPattern() {
        String pattern = "timestamp";
        DateParser parser = new DateParser(pattern);
        ByteBuffer bb = SimpleDateType.instance.fromString("2015-10-10");
        Column<Integer> column = Column.builder("date").decomposedValue(bb, SimpleDateType.instance);
        Date actualDate = parser.parse(column);
        Date expectedDate = parser.parse(1444435200000L);
        assertEquals("Date parser fails with SimpleDateType and timestamp pattern", expectedDate, actualDate);
    }

}

File: plugin/src/test/scala/com/stratio/cassandra/lucene/common/DateParserTest.scala
/*
 * Copyright (C) 2014 Stratio (http://stratio.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.stratio.cassandra.lucene.common

import java.text.DateFormat
import java.util.Date

import com.stratio.cassandra.lucene.BaseScalaTest
import org.apache.cassandra.utils.UUIDGen
import org.junit.runner.RunWith
import org.scalatest.junit.JUnitRunner
/**
  * @author Eduardo Alonso  `eduardoalonso@stratio.com`
  */
@RunWith(classOf[JUnitRunner])
class DateParserTest extends BaseScalaTest {
  test("parse null") {
    BaseScalaTest.assertNull("yyyy/MM/dd", null)
  }

  test("parse date ") {
    val date = BaseScalaTest.date("yyyy/MM/dd", "2015/11/03")
    BaseScalaTest.assertEquals("yyyy/MM/dd", date, date)
  }

  test("parse min date") {
    val date = new Date(0)
    BaseScalaTest.assertEquals(DateParser.DEFAULT_PATTERN, date, date)
  }

  test("parse max date") {
    val date = new Date(Long.MaxValue)
    BaseScalaTest.assertEquals(DateParser.DEFAULT_PATTERN, date, date)
  }


  test("parse date truncating") {
    val date = BaseScalaTest.date("yyyy/MM/dd HH:mm:ss", "2015/11/03 01:02:03")
    val  expected = BaseScalaTest.date("yyyy/MM/dd", "2015/11/03")
    BaseScalaTest.assertEquals("yyyy/MM/dd", date, expected)
  }

  test("parse integer") {
    val date = BaseScalaTest.date("yyyyMMdd", "20151103")
    BaseScalaTest.assertEquals("yyyyMMdd", 20151103, date)
  }

  test("parse invalid integer") {
    BaseScalaTest.assertFail("yyyyMMdd", 1)
  }

  test("parse negative integer") {
    BaseScalaTest.assertFail("yyyyMMdd", -20151103)
  }

  test("parse long") {
    val date = BaseScalaTest.date("yyyyMMdd", "20151103")
    BaseScalaTest.assertEquals("yyyyMMdd", 20151103L, date)
  }

  test("parse invalid long") {
    BaseScalaTest.assertFail("yyyyMMdd", 1L)
  }

  test("parse negative long ") {
    BaseScalaTest.assertFail("yyyyMMdd", -20151103L)
  }

  test("parse float") {
    val date = BaseScalaTest.date("yyyy", "2015")
    BaseScalaTest.assertEquals("yyyy", 2015f, date)
  }

  test("parse decimal float") {
    val date = BaseScalaTest.date("yyyy", "2015")
    BaseScalaTest.assertEquals("yyyy", 2015.7f, date)
  }

  test("parse invalid float") {
    BaseScalaTest.assertFail("yyMM", 1f)
  }

  test("parse negative float") {
    BaseScalaTest.assertFail("yyyy", -2015f)
  }

  test("parse double") {
    val date = BaseScalaTest.date("yyyyMM", "201511")
    BaseScalaTest.assertEquals("yyyyMM", 201511d, date)
  }

  test("parse decimal double") {
    val date = BaseScalaTest.date("yyyyMM", "201511")
    BaseScalaTest.assertEquals("yyyyMM", 201511.7d, date)
  }

  test("parse invalid double") {
    BaseScalaTest.assertFail("yyyyMM", 1d)
  }

  test("parse negative double") {
    BaseScalaTest.assertFail("yyyyMM", -201511d)
  }

  test("parse string") {
    val expected = BaseScalaTest.date("yyyy/MM/dd", "2015/11/03")
    BaseScalaTest.assertEquals("yyyy/MM/dd", "2015/11/03", expected)
  }
  test("parse invalid string") {
    BaseScalaTest.assertFail("yyyy/MM/dd", "20151103")
  }

  test("parse uuid") {
    val uuid = UUIDGen.getTimeUUID(BaseScalaTest.date("yyyy-MM-dd HH:mm", "2015-11-03 06:23").getTime)
    val expected = BaseScalaTest.date("yyyyMMdd", "20151103")
    BaseScalaTest.assertEquals("yyyyMMdd", uuid, expected)
  }
}
